source/fe/mapping_q_generic.cc

```diff
@@ -2286,23 +2287,34 @@ namespace internal
                               *
                               data.mapping_support_points[k][i]);
 
-                  // pushing forward the derivative coordinates
+                  // first push forward the j-components
                   for (unsigned int i=0; i<spacedim; ++i)
                     for (unsigned int j=0; j<spacedim; ++j)
-                      for (unsigned int l=0; l<spacedim; ++l)
-                        {
-                          jacobian_pushed_forward_grads[point][i][j][l] = result[i][0][0] *
-                                                                          data.covariant[point][j][0] *
-                                                                          data.covariant[point][l][0];
-                          for (unsigned int jr=0; jr<dim; ++jr)
-                            {
-                              const unsigned int lr_start = jr==0? 1:0;
-                              for (unsigned int lr=lr_start; lr<dim; ++lr)
-                                jacobian_pushed_forward_grads[point][i][j][l] += result[i][jr][lr] *
-                                                                                 data.covariant[point][j][jr] *
-                                                                                 data.covariant[point][l][lr];
-                            }
-                        }
+                        for (unsigned int l=0; l<dim; ++l)
+                          {
+                            tmp[i][j][l] = result[i][0][l] *
+                                           data.covariant[point][j][0];
+                            for (unsigned int jr=1; jr<dim; ++jr)
+                              {
+                                tmp[i][j][l] += result[i][jr][l] *
+                                                data.covariant[point][j][jr];
+                              }
+                          }
```

Comment:
That's a good argument. Thanks for looking into it. I guess we let the compiler find out if there are possibilities for data reuse.